### PR TITLE
2023-11-22 stable API (porting 07-01-preview api)

### PR DESCRIPTION
**Pull request by jhoreman**
### Which issue this PR addresses:

https://issues.redhat.com/browse/ARO-4613

### What this PR does / why we need it:

This ports functionality from the latest preview API into a stable API definition. 

### Test plan for issue:

<!--
How did you test that this PR works?

Need to deploy a cluster through the API running locally. 

### Is there any documentation that needs to be updated for this PR?

No

### CODE REVIEW

File: pkg/util/azureclient/mgmt/redhatopenshift/2023-11-22/redhatopenshift/openshiftclusters.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+// NewOpenShiftClustersClient creates a new OpenShiftClustersClient
+func NewOpenShiftClustersClient(environment *azureclient.AROEnvironment, subscriptionID string, authorizer autorest.Authorizer) OpenShiftClustersClient {
+	var client mgmtredhatopenshift20231122.OpenShiftClustersClient
+	if env.IsLocalDevelopmentMode() {
+		client = mgmtredhatopenshift20231122.NewOpenShiftClustersClientWithBaseURI("https://localhost:8443", subscriptionID)
+		client.Sender = &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: true, // #nosec G402
```
[the remark this answers is not in the document]
Comment (jhoreman) [reply pkg/util/azureclient/mgmt/redhatopenshift/2023-11-22/redhatopenshift/openshiftclusters.go:40]: I believe we can accept and dismiss this check as it is only relevant in the context of a local RP call.

File: pkg/util/azureclient/mgmt/redhatopenshift/2023-11-22/redhatopenshift/operations.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+// NewOperationsClient creates a new OperationsClient
+func NewOperationsClient(environment *azureclient.AROEnvironment, subscriptionID string, authorizer autorest.Authorizer) OperationsClient {
+	var client mgmtredhatopenshift20231122.OperationsClient
+	if env.IsLocalDevelopmentMode() {
+		client = mgmtredhatopenshift20231122.NewOperationsClientWithBaseURI("https://localhost:8443", subscriptionID)
+		client.Sender = &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: true, // #nosec G402
```
[the remark this answers is not in the document]
Comment (jhoreman) [reply pkg/util/azureclient/mgmt/redhatopenshift/2023-11-22/redhatopenshift/operations.go:36]: I believe we can accept and dismiss this check as it is only relevant in the context of a local RP call.

File: pkg/api/v20231122/openshiftcluster.go [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+	// The desired managed outbound IPs for the cluster public load balancer.
+	ManagedOutboundIPs *ManagedOutboundIPs `json:"managedOutboundIps,omitempty" mutable:"true"`
+	// The list of effective outbound IP addresses of the public load balancer.
+	EffectiveOutboundIPs []EffectiveOutboundIP `json:"effectiveOutboundIps,omitempty" swagger:"readOnly"`
+	// The desired outbound IP resources for the cluster load balancer.
+	OutboundIPs []OutboundIP `json:"outboundIps,omitempty" mutable:"true"`
+	// The desired outbound IP Prefix resources for the cluster load balancer.
+	OutboundIPPrefixes []OutboundIPPrefix `json:"outboundIpPrefixes,omitempty" mutable:"true"`
+	// The desired number of allocated SNAT ports per VM. Allowed values are in the range of 0 to 64000 (inclusive). The default value is 1024.
+	AllocatedOutboundPorts *int `json:"allocatedOutboundPorts,omitempty" mutable:"true"`
```
Comment (tony-schndr) [inline pkg/api/v20231122/openshiftcluster.go:160]: phase 2/3 fields

Comment (jhoreman) [reply pkg/api/v20231122/openshiftcluster.go]: Removed in latest commit

**Review by SudoBrendan — COMMENTED** [review]
To future reviewers - this PR includes both the generation of the skeleton API and adding the new fields this API supports. To view what is changing here, you can run:

```
diff -bur v20230904/ v20231122/
```

File: pkg/api/v20231122/openshiftcluster_validatestatic_test.go [hunk 4]
```diff
@@ -0,0 +1,1004 @@
...
+			wantErr: "400: InvalidParameter: properties.networkProfile.serviceCidr: The provided vnet CIDR '10.0.0.0/23' is invalid: must be /22 or larger.",
+		},
+		{
+			name: "OutboundType is empty",
+			current: func(oc *OpenShiftCluster) {
+				oc.Properties.NetworkProfile.OutboundType = ""
+			},
+			wantErr: "",
+		},
+		{
```
Comment (SudoBrendan) [inline pkg/api/v20231122/openshiftcluster_validatestatic_test.go:519]: This file has a lot of diffs compared to the preview api (missing tests in this func, missing func `TestOpenShiftClusterStaticValidateLoadBalancerProfile`, and other missing tests too unrelated to the phase 2/3 fields). Can you resolve the diff?

**Review by SudoBrendan — CHANGES_REQUESTED** [review]
There appears to be a diff in one file compared to the preview API that's relevant.

File: pkg/api/v20231122/openshiftcluster_validatestatic_test.go (hunk 4 above)
Comment (jhoreman) [reply pkg/api/v20231122/openshiftcluster_validatestatic_test.go:519]: Delta should match now.

File: pkg/api/v20231122/openshiftcluster_example.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	return oc
+}
+
+// ExampleOpenShiftClusterResponse returns an example OpenShiftCluster object
+// that the RP might return to an end-user in a GET response
+func ExampleOpenShiftClusterGetResponse() interface{} {
+	oc := exampleOpenShiftCluster()
+	oc.Properties.ClusterProfile.PullSecret = ""
+	oc.Properties.ServicePrincipalProfile.ClientSecret = ""
+	oc.Properties.NetworkProfile.LoadBalancerProfile = &LoadBalancerProfile{
```
Comment (SudoBrendan) [inline pkg/api/v20231122/openshiftcluster_example.go:57]: diff: missing `oc.Properties.WorkerProfilesStatus = nil` that's in stable.

File: pkg/api/v20231122/openshiftcluster_convert.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+		LastModifiedBy:     oc.SystemData.LastModifiedBy,
+		LastModifiedAt:     oc.SystemData.LastModifiedAt,
+		LastModifiedByType: api.CreatedByType(oc.SystemData.CreatedByType),
+	}
+}
+
+// ExternalNoReadOnly removes all read-only fields from the external representation.
+func (c openShiftClusterConverter) ExternalNoReadOnly(_oc interface{}) {
+	oc := _oc.(*OpenShiftCluster)
+	if oc.Properties.NetworkProfile.LoadBalancerProfile != nil {
```
Comment (SudoBrendan) [inline pkg/api/v20231122/openshiftcluster_convert.go:266]: diff: missing `oc.Properties.WorkerProfilesStatus = nil` that's in stable.

File: pkg/api/v20231122/openshiftcluster_validatestatic.go [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+	}
+
+	return sv.validateProperties("properties", &oc.Properties, isCreate)
+}
+
+func (sv openShiftClusterStaticValidator) validateProperties(path string, p *OpenShiftClusterProperties, isCreate bool) error {
+	switch p.ProvisioningState {
+	case ProvisioningStateCreating, ProvisioningStateUpdating,
+		ProvisioningStateAdminUpdating, ProvisioningStateDeleting,
+		ProvisioningStateSucceeded, ProvisioningStateFailed:
```
Comment (SudoBrendan) [inline pkg/api/v20231122/openshiftcluster_validatestatic.go:85]: diff: missing `ProvisioningStateCancelled` that's in preview.

**Review by SudoBrendan — CHANGES_REQUESTED** [review]
A few more diffs that need review.

File: pkg/api/v20231122/openshiftcluster_validatestatic.go (hunk 7 above)
Comment (jhoreman) [reply pkg/api/v20231122/openshiftcluster_validatestatic.go]: Added, thanks.

File: pkg/api/v20231122/openshiftcluster_convert.go (hunk 6 above)
Comment (jhoreman) [reply pkg/api/v20231122/openshiftcluster_convert.go]: Added, thanks.

File: pkg/api/v20231122/openshiftcluster_example.go (hunk 5 above)
Comment (jhoreman) [reply pkg/api/v20231122/openshiftcluster_example.go]: Added, thanks.

**Comment by tony-schndr** [comment]
LGTM, I went through the diffs it looks like everything is here.

**Review by SudoBrendan — APPROVED** [review]

**Pull request merged**